### PR TITLE
#9753 - Shadow export csv file contains an extra heading field that causes importing back crashing

### LINKED ISSUE #9753: Shadow export csv file contains an extra heading field that causes importing back crashing
Issue overview
--------------
It was discovered that (at least in some scenarios?) that the shadow calculation export .csv file will/could produce an extra ",()" field at the end of the first line (the headings line): 
image

This will cause EnergyPlus to crash when importing this .csv file back using the "imported" shading calculation method. The processing code will encounter some out of bound error when reading the column and causing a crash.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): Windows, possibly all.
 - Version of EnergyPlus (if using an intermediate build, include SHA): v9.6 to v22.2
 - Helpdesk ticket number: discovered while investigating Helpdesk Ticket 16292.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9753))
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9753
 - When exporting, remove these parenthesis / extra header field
 - When importing, if there is this extra one, don't crash but remove the entry (backward compat)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 1]
```diff
@@ -1451,3 +1456,110 @@ TEST_F(EnergyPlusFixture, ScheduleFileDSTtoggleOptionTest)
...
+    ASSERT_TRUE(process_idf(idf_objects));
+
+    state->dataGlobal->NumOfTimeStepInHour = 4; // must initialize this to get schedules initialized
+    state->dataGlobal->MinutesPerTimeStep = 15; // must initialize this to get schedules initialized
+    state->dataGlobal->TimeStepZone = 0.25;
+    state->dataGlobal->TimeStepZoneSec = state->dataGlobal->TimeStepZone * DataGlobalConstants::SecInHour;
+    state->dataEnvrn->CurrentYearIsLeapYear = false;
+
+    EXPECT_FALSE(state->dataScheduleMgr->ScheduleFileShadingProcessed);
+    ScheduleManager::ProcessScheduleInput(*state); // read schedules
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1480]: Backward compat test. Without backward compat fix, it segfaults on the ProcessScheduleInput

File: src/EnergyPlus/ScheduleManager.cc [hunk 2]
```diff
@@ -505,6 +505,18 @@ namespace ScheduleManager {
...
+                ShowWarningError(state,
+                                 format("{}{}=\"{}\" Removing last column of the CSV since it has '()' for the surface name.",
+                                        RoutineName,
+                                        CurrentModuleObject,
+                                        Alphas(1)));
+                ShowContinueError(state, "This was a problem in E+ 22.2.0 and below, consider removing it from the file to suppress this warning.");
+                schedule_file_shading_result->second["header"].erase(NumCSVAllColumnsSchedules);
+                schedule_file_shading_result->second["values"].erase(NumCSVAllColumnsSchedules);
+                --NumCSVAllColumnsSchedules;
+            }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:518]: Backward compat fix + issue a warning in eplusout.err when reading a Schedule:File:Shading file that has an extra `()` header

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 3]
```diff
@@ -3902,3 +3903,76 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_Warn_Pixel_Count_and_TM_Schedule)
...
+    WindowManager::initWindowModel(*state);
+
+    state->dataSolarShading->GetInputFlag = false;
+    HeatBalanceManager::InitHeatBalance(*state);
+
+    {
+        auto const stream_str = state->files.shade.get_output();
+        EXPECT_EQ(expected_values, stream_str);
+    }
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarShading.unit.cc]: Export test.

File: src/EnergyPlus/HeatBalanceManager.cc [hunk 4]
```diff
@@ -6627,7 +6627,7 @@ namespace HeatBalanceManager {
         for (SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; ++SurfNum) {
             print(state.files.shade, "{},", state.dataSurface->Surface(SurfNum).Name);
         }
-        print(state.files.shade, "()\n");
+        print(state.files.shade, "\n");
```
Comment (jmarrec) [inline src/EnergyPlus/HeatBalanceManager.cc:6630]: The fix for the export is simple, remove these parentheses

Comment (Myoldmopar) [reply src/EnergyPlus/HeatBalanceManager.cc:6630]: Super simple and straightforward.  Trying to trace this change back became tedious and I gave up.  Anyway, it's an obvious fix, thanks.

File: src/EnergyPlus/ScheduleManager.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ScheduleManager.cc:518]: This is a reasonable warning message and cleans up the data to remove the erroneous value.  :+1:

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc (hunk 1 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/ScheduleManager.unit.cc:1480]: :+1:

**Review by Myoldmopar — APPROVED** [review]
Looks good to me!

**Comment by Myoldmopar** [comment]
This all tests real happy locally and is an obvious fix.  I'm not going to wait on more CI cycles for this.  Thanks @jmarrec

File: src/EnergyPlus/HeatBalanceManager.cc (hunk 4 above)
Comment (jmarrec) [reply src/EnergyPlus/HeatBalanceManager.cc:6630]: It was there in the gio days, except there was a refactor there too.

It was broken starting at 9.1.0, It's this commit: https://github.com/NREL/EnergyPlus/commit/39ffc4329428e1117dc50057efdb1b9f85f8f8a5

**Pull request merged**